src/2025h1/unsafe-fields.md

```diff
@@ -0,0 +1,114 @@
+# Unsafe Fields
+
+| Metadata |          |
+| ---      | ---      |
+| Owner(s) | @jswrenn |
+| Teams    | [lang]   |
+| Status   | Proposed |
+
+## Summary
+
+Design and implement a mechanism for denoting when fields carry library safety invariants.
+
+## Motivation
+
+The absence of a mechanism for denoting the presence of library safety invariants increases both the risk of working with `unsafe` code and the difficulty of evaluating its soundness.
+
+### The status quo
+
+Presently, Rust lacks mechanisms for denoting when fields carry library safety invariants, and for enforcing extra care around their use. Consequently, to evaluate the soundness of `unsafe` code (i.e., code which relies on safety invariants being upheld), it is not enough to check the contents of `unsafe` blocks — one must check all places (including safe contexts) in which safety invariants might be violated. (See [*The Scope of Unsafe*])
+
+For example, consider this idealized `Vec`:
+
+```rust
+pub struct Vec<T> {
+    data: Box<[MaybeUninit<T>]>,
+    len: usize,
+}
+```
+
+Although `len` is bound by a safety invariant, it is trivial to violate its invariant in entirely safe code:
+
+```rust
+impl Vec<T> {
+    pub fn evil(&mut self) {
+        self.len += 2;
+    }
+}
+```
+
+Rust cannot enforce that modifications of `len` require `unsafe`, because the language does not provide the programmer a way of communicating to the compiler that `len` carries safety invariants.
+
+[*The Scope of Unsafe*]: https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html
+
+### The "shiny future" we are working towards
+
+Rust programmers will use the `unsafe` keyword to denote fields that carry library safety invariants; e.g.:
+
+```rust
+struct Vec<T> {
+    // SAFETY: The elements `data[i]` for
+    // `i < len` are in a valid state.
+    unsafe data: Box<[MaybeUninit<T>]>,
+    unsafe len: usize,
+}
+```
+
+Rust will require that usages of `unsafe` fields which could violate their safety invariants must *only* occur within `unsafe` contexts.
+
+### The next 6 months
+
+In the next six months, we will iterate on the design and implementation of unsafe fields. An RFC for `unsafe` fields will be accepted, and a candidate implementation will — at the very least — be ready to enter the stabilization process.
+
+## Design axioms
+
+The design of `unsafe` fields is guided by three axioms:
+
+1. **Unsafe Fields Denote Safety Invariants**
+   A field *should* be marked `unsafe` if it carries arbitrary library safety invariants with respect to its enclosing type.
+2. **Unsafe Usage is Always Unsafe**
+   Uses of `unsafe` fields which could violate their invariants *must* occur in the scope of an `unsafe` block.
+3. **Safe Usage is Usually Safe**
+   Uses of `unsafe` fields which cannot violate their invariants *should not* require an unsafe block.
```

Comment:
What does this mean, in practice?

Reply:
For example: Using [`offset_of`](https://doc.rust-lang.org/nightly/core/mem/macro.offset_of.html) has no impact on the safety invariants of its given field. Requiring `unsafe { ... }` to compute the offset of an unsafe field would be pure noise in an unsafe code review, and, as such, we should make an effort to *not* require `unsafe { ... }` in this case.

Unlike the first and second tenet, this tenet is a soft requirement. We shouldn't ignore it entirely in our design work, but we don't actually have to implement any of these safe carve-outs initially since it's forwards-compatible to make an unsafe operation safe (so long as doing so doesn't violate existing safety conditions).